### DISCUSSION TITLE
新功能讨论（New feature discussion）

### DISCUSSION THREAD

**Discussion by LincZero** [any-block/any-block#14] [General]
Refactor selector：Not compatible with older versions
- Removed （Reasons for incompatibility with older versions）
    - Removed the "always in effect" setting for the selector and now cannot use the selector without header information
    - The old head-tail selector has been removed and replaced with a VuePress syntax head-tail selector（`:::`）
- Add
    - **[Major Update] v2.1.0 allows the use of selectors within nested lists, references, and indents**
    - **[Major Update] v2.2.0 allows selectors to nest selectors**
    - Add an English version of the manual
    - The Settings panel was refactored and a lot of Settings were removed
- Fixing
    - #6 #9 #7
- Known bug
    - There are more bugs in render mode (read mode)

---

项目重构了：不与旧版本兼容
- 删除项（不兼容旧版本的原因）
    - 删除了选择器 “总是生效” 的设置，现在无法在没有头部信息的情况下使用选择器
    - 删除了旧的首尾选择器，换成了VuePress语法的首尾选择器（`:::`）
- 增加项
    - **【主要更新】v2.1.0 允许在列表、引用、缩进的嵌套中使用选择器**
    - **【主要更新】v2.2.0 允许选择器套选择器**
    - 增添英文版的手册
    - 重构了设置面板，删去了大量设置
    - 表格选择器
- 修复项
    - #6 #9 #7
- 已知bug
    - 渲染模式（阅读模式）的bug增多了

<hr /><em>This discussion was created from the release <a href='https://github.com/LincZero/obsidian-any-block/releases/tag/2.2.0'>2.2.0</a>.</em>

**Comment by LincZero** [comment]
![图片](https://user-images.githubusercontent.com/25877214/223704064-29d6a49f-8f60-43ab-a38a-b077bb5c9974.png)
我构思了一个AnyBlock的新处理器，语法是这样的

**Reply by LincZero** [reply]
### parameter

**[floder-]**
折叠至根层（完全不展开）
Fold to root layer (no expansion at all)

**[floder+]**
默认文件夹完全展开
The default folder is fully expanded

**[floder]**
（默认）同 `floader-`（还是同 `floder+` 会好一点？）
（Default）Same as `floader-` (or is `floder+` better?)

**[floader--]**
仅展开至两层（几个减号就展开至几层）
Expand to two levels only (several minus signs expand to several levels)

**Reply by LincZero** [reply]
想过选择在无anyblock插件下体验较好的语法：

0. 目前的方案`<`+空格：缺点：和之前的指定有无表头的语法是一样，会冲突
1. 复选框：样式不错，但有歧义，有点怪（你不小心按到复选框就更怪了）
2. 列表嵌套标题：这里仅建议4级标题及以上，优点：无ab插件下的渲染体验好
    缺点：
    - 样式污染，列表项会变大变小，不符合文件夹的逻辑（虽然可以用css解决）
    - 就算用css解决，但用户可能会用到列表嵌标题，不能和他们的用法冲突
    - 不建议用户用4级以下标题，但那样的话要敲的键可能会增多？除非快捷键
3. 列表嵌套引用。缺点：样式差，还不如嵌套标题呢
4. 文件夹的列表项加粗：加粗太麻烦了，而且说不定用户可能也需要加粗这个功能呢
    优点：无ab插件下的渲染体验好

你们看有什么好想法不

**Reply by LincZero** [reply]
与目前的list2lt同模，就是在2lt的基础上：

1. 删除了表格边框并优化样式
2. 加入了区分文件夹/文件的图标
（除了文件或文件夹外区分外，是否会根据文件后缀名再给不同文件类型不同的图标样式……这点我暂时不考虑，先把这个实现了再说，有需要的话以后再提issue）
3. 提供了默认展开层级的可选参数

所以你可以轻易地给文件列表折叠，以及加上多层 “文件/文件夹备注”

**Reply by Linzeal** [reply]
> ### parameter
> **[floder-]** 折叠至根层（完全不展开） Fold to root layer (no expansion at all)
> 
> **[floder+]** 默认文件夹完全展开 The default folder is fully expanded
> 
> **[floder]** （默认）同 `floader-`（还是同 `floder+` 会好一点？） （Default）Same as `floader-` (or is `floder+` better?)
> 
> **[floader--]** 仅展开至两层（几个减号就展开至几层） Expand to two levels only (several minus signs expand to several levels)

[floader--] 仅展开至两层（几个减号就展开至几层）不如使用 [floader2] 或 [floader(2)] 更简单明了，数字几就是展开至几层。

**Comment by LincZero** [comment]
I have conceived of a new processor for AnyBlock with this syntax.
Any ideas I have until the next update will be here, okay.
In addition to raising a more formal issue, you can also raise some ideas here.
————————
到下版本更新之前，我有什么想法都会写在这。
你们除了提比较正式的issue，也可以在这里提点想法。

**Comment by LincZero** [comment]
![图片](https://user-images.githubusercontent.com/25877214/224003323-03a3c801-ef68-455c-b222-c70fb0c3e8ae.png)

你是否有过这种需求：想给表格的前N个行和后N个行做一个小的分隔？你可能会使用空的表格行？或者新增一行内容为 `---------` 的表格行？

功能设计：多段表格 / 表格切割线 的支持（不增加新处理器，而是让原来的 `2table、2lt、2ut` 处理器支持该功能），新语法拟定与效果设计如图。

---

Have you ever wanted to make a small separation between the first N rows and the last N rows in a table? You might use an empty table row? Or add a new table row with '---------'?

Function design: Support for multi-segment table/table cutting line (no new processor is added, but the original `2table, 2lt, 2ut` processors support this function), new syntax and effect design are shown in the figure.

**Comment by LincZero** [comment]
灵感：
对折叠[fold]处理器做一个升级：[foldthis]。
新处理器的选择器为非列表了标题时，和旧版本效果一样。
新处理器的选择器为列表或标题时，无视选择器的选择范围，通过js影响下面的列表或标题，将其设置为折叠状态。

好处：使用了原生的折叠系统，更简单。
缺点：不是对选择器区域进行一次包装再折叠，这意味着该处理器所影响的范围可能不与选择器相同。
这会导致这个处理器是“特殊”的，==会引发歧义==，这是最严重的问题。

缺点，例如：
[foldthis]
- 这里会折叠上
    - 1
    - 2
- 而这里并不会被折叠上
    - 1
    - 2

当然也列表二也都折叠上，但和原来的 `fold` 选择器还是有所区别的，无法将其完全折叠。
总体来说，这个主要是给标题而非列表用的。作用是将**可折叠**转化为**默认折叠**，这是两个概念。

引发歧义是最严重的问题，可能会考虑的修复方案：`特殊选择器`（区分于自动选择器，这个以后再说，暂时不打算做出来）

**Comment by Linzeal** [comment]
折叠这个想法不错，有这功能后可以通过自动折叠来显示出重点。这个也可以用在[list2lt]可折叠表格上。

就是`<`字符的用途不太明白，是用来表示本层级为默认自动折叠状态的标志吗？如果是这个作用，感觉字符`<`会与表示表头相混淆，或许直接用`>`字符来表示本层级默认折叠可展开会更好，字符`>`与Obsidian里表示折叠状态的图标是相近的，用户熟悉其代表的意思。

**Reply by LincZero** [reply]
`>` 字符是引用啊，列表嵌引用的样式我上面的评论说过这个问题，样式太怪了。

我现在的想法是参考插件 `list callout`，其实可以用的字符非常多，Folder处理器中甚至可以做到Linux终端里显示文件/文件夹的那种效果（绿字 蓝字 白字 或者绿底 蓝底等）

但主要是这个符号没意义，加多了怕用户乱用引起混淆。目前首选的符号是 `=`和`~`，分别代表文件夹图标和根据文件后缀选择图标（不加的话就是不显示图标）

**Comment by Linzeal** [comment]
我想了一下，AnyBlock插件的主旨是将任何东西（ANY）转化为某种特定样式的块（BLOCK）。总体可分为对文字样式的转化渲染和计算的转化渲染。
目前，不管是list2table，还是addClass等等之类，都是对文本样式的转化渲染，这是一个方向大类。
而计算方面的转化渲染还欠缺，我认为哪怕AB自己不搞，也应当有这个接口API，这样AB才能称得上名称中的ANY。

计算转化，我想了两个使用场景，一个是集中记账式，这也是很多Obsidian用户的需求之一，可参考转化为Excel表格来计算，如截图中例1：
可用缩进列表形式体现数值对，计算表达式左边（一般有`=，`或空格等字符来分隔）为变量或为空，计算表达式右边（一般有`，` `//` 或空格等字符来分隔）为备注或为空。整体渲染为Excel表格样式，分为三列，最左为变量，中间为计算表达式，最右为备注。

另一个是行文中的计算，如截图中例2：
以代码符形式体现变量值对，同样统一的计算表达式左边（一般有`=，`或空格等字符来分隔）为变量或为空，计算表达式右边（一般有`，` `//` 或空格等字符来分隔）为备注或为空

![image](https://user-images.githubusercontent.com/20738431/226153704-47b6ef57-3ae1-496a-99b8-041cccb7a038.png)

**Reply by LincZero** [reply]
帐表需求我在群里也见有人提过……但有点麻烦且与本人需求不吻合，不考虑，API我尽量完善封装和文档，再看有谁愿意做吧。

自定义分隔符我倒是有可能会做，让用户使用 `|  ` 符号之外的分隔符。而且有一个比较通用的解决方案：
使用文本处理器，将自定义符号转化为 `| `，例如：

设置别名：`[separator(, )]` alias `[replace(", ","| ")]`
`replace` 就是一个简单的替换处理器，当然他不会真实修改你的源文件，只是处理渲染出来的文本

**Reply by Linzeal** [reply]
有API就好。

另外，发现你还弄了一套英文README和demo，感觉没必要弄两套README和demo，README要分README.md和README_zh两个，demo也分demo和demo_zh两份，连截图都要中文英文两份，这样在两套文件中重复修改，又累又麻烦不说，其实是没必要。

中文是未来，一份README和demo以中文为主体即可，有闲暇精力在每一段中文段落后面或下边给出对应的英文翻译就行，更改中文段落时当即就就能同时更改英文。若是中英分两套，中文改了，还要找到英文那里再改，麻烦。

**Comment by Linzeal** [comment]
还有一个与 [2tab] 很相似的功能：分栏。而且用户需求也很大，尤其写稿子的。

**Comment by Linzeal** [comment]
还有一个功能，目录列表 [TOC]，这在对外发布文章时总是需要。很多Markdown编辑器有内置 [TOC] 功能，但Obsidian没有内置，只有一个大纲面板。

[TOC] 应当要能控制目录的显示级别，比如，不带参数的 [TOC] 默认显示所有级别的目录，带参数的则参考 [folder] 的折叠功能，如 [TOC2] 或 [TOC(2)] 默认将3级以上的目录自动折叠，仅显示2级以内的目录。

**Reply by LincZero** [reply]
功能其实不重合，不会放在这个插件里。

这个ab总体来说就是一个 “选择范围” 然后 “处理器处理该范围内的文本” 的逻辑。

无论是底层实现还是表面的逻辑都完全不同，不适合放在这个插件里。

**Reply by Linzeal** [reply]
> 功能其实不重合，不会放在这个插件里。
> 
> 这个ab总体来说就是一个 “选择范围” 然后 “处理器处理该范围内的文本” 的逻辑。
> 
> 无论是底层实现还是表面的逻辑都完全不同，不适合放在这个插件里。



嗯，也就是说AB只是选定后并将其进行转化。

那看来 [TOC] 是不太合适。现在我能想到的其他符合选定并转化功能的就下面这些了：

1. 我前面提到的分栏 [2columns] 功能应该符合。

2. 把计算表达式转化为计算结果 [2math] 这个功能应该也是符合，借助 math.js 库能省很多事。

3. 再来一个功能 [2QR]，就是转化为QR二维码，也可借助GitHub里已有的QR库。比如把链接转化QR，或者把一段文本转为QR。这样，在移动端扫一扫就可取得内容。

**Reply by LincZero** [reply]
手动的分栏其实可以用 [2ut|addClass(无边框)]，像BT库的分栏实现其实就是纯css修改出来的样式。
像word那种自动分栏的话……如果掺杂了多种样式，不是纯文本的话，也做不了（很难去计算行距和边距，你又不知道他们用的什么主题）

最近都是在等上架再加快更新频率，最近都没怎么补充这个插件的功能，没上架没动力。倒是最近在研究 github page 和 vuepress，下周会有github page网站来作为单独的 anyblock 使用手册。会方便很多，顺便国际化（多语言）起来会比较方便

**Comment by LincZero** [comment]
插件审核了两个月余一周终于上架，恢复更新 !!!

目前的首要todo优先级顺序：

1. 重构代码
    - 使用泛型重构处理器接口（已完成大半）
    - 基于上一点，扩展三种处理器类别（原来只有text和element 2种）：tree-node、list-line、table-cell
    - 将之前的一些列表和表格函数修改成可复用的处理器
2. 增加必要的处理器
    - replace处理器（替换文本）
    - 分级别归纳的list处理器（例如只收纳一层、只收纳二层等）
      非常有用，由此基础上扩展标题处理器、时间线、文件树
    - 增加El反向处理器（将El类型反向生成html文本）
3. 独立插件文档网站（之前事情多被鸽掉了）
4. 特殊处理器 - ab空处理器与严格模式
5. 特殊选择器 - 原生折叠
6. 基于anyblock的markmap插件的开发，顺便成为用来演示如何基于ab进行插件扩展的demo插件

**Comment by laogao2005** [comment]
@LincZero 非常棒的一个插件！使用中有几个问题请教：
①为什么我们的timeline这么难看，而你教程里的清爽极了？我用的是blue topaz主题
![11](https://github.com/user-attachments/assets/f758c9f6-1bf3-4ba3-919e-128983c12b42)
②请问如何实现tab表头内，嵌入目录或其他的block，如你下面的图示效果？
![22](https://github.com/user-attachments/assets/22ce78c4-f5e0-4146-94c3-486cea3899d3)

谢谢！

**Comment by LincZero** [comment]
（1）

点只有一半那个确实有问题，对应修复：

```
.ab-note table.ab-table-timeline tr:first-child td {
  overflow: visible; /* 修复个别情况下那个点被截断 */
}
```

这个宽度问题，看了下实时模式没有这个问题，阅读模式存在。被BT主题的一个css给影响了。消除影响：（不确定会不会误影响其他）

```
:is(.markdown-preview-view,.markdown-rendered).is-readable-line-width:not(.matrix) :is(.markdown-rendered) {
    width: auto;
}
```

BlueTopaz 主题样式太多太容易误影响其他东西了，比较少去适配，适配压力比较大。

上面两个样式你可以在设置 > 外观 > CSS代码片段先自行添加解决，我这刚更新完没那么快发下版本

（2）

建议先读：https://linczero.github.io/MdNote_Public/ProductDoc/AnyBlock/docs/zh/03.%20%E9%80%89%E6%8B%A9%E5%99%A8.html

列表选择器套列表选择器（不建议，导致该列表非常臃肿。不建议列表再去嵌套其他东西）

```md
[tabs]

- 11
  [tabs]
  - 21
    210
  - 22
    220
  - 23
    230
- 12
  120
- 13
  130
```

标题选择器嵌套其他选择器（一般外面用标题选择器或mdit选择器会比较好）

```md
[tabs]

## 11

11000

[tabs]
- 21
  210
- 22
  220
- 23
  230

## 12

120
## 13

130
```

mdit选择器套其他选择器

```md
::: tabs

@tab 1

……

[tabs]

- x
  x0
- y
  y0

@tab 2

20

@tab 3

30

:::
```

**Reply by laogao2005** [reply]
好的，非常感谢！

**Comment by laogao2005** [comment]
@LincZero 您好，请问，选择器里放入button是不允许的，对吗？
![Quicker_20240926_001122](https://github.com/user-attachments/assets/c6715378-0201-4d5b-95cd-a88d3f090e21)


[tabs]
- 标签表1
 [卡片]
  -  `button-FB-ZBWYH`
  - Card2
  - card3
- 标签表2
 [卡片]
  - fdddgjk
  - gfhhgf
- 标签表3
 [目录]
  - ghdf/
	  - GHJF
  - hjio/
	  - DGGF

**Reply by LincZero** [reply]
应该是可以的，这是什么插件？

要求与之配合的插件能在阅读模式下生效，也能在列表/引用块这种嵌套内生效。一般满足这两的应该都可以的

**Reply by laogao2005** [reply]
> 应该是可以的，这是什么插件？
> 
> 要求与之配合的插件能在阅读模式下生效，也能在列表/引用块这种嵌套内生效。一般满足这两的应该都可以的

re：插件就是[buttons](https://github.com/shabegom/buttons) 和anyblock呀

**Reply by LincZero** [reply]
下次这一类的在issue那提问题，别在discussions。

看了下，似乎是ob的重渲染，会导致button获取file值失败，进而报错。估计有可能要他那边修

![图片](https://github.com/user-attachments/assets/db9d4b6d-a14f-484c-8ef2-4e68ae54fd51)

**Reply by LincZero** [reply]
看了下buttons插件的issue，和这个是同一个问题：

https://github.com/shabegom/buttons/issues/212

**Reply by LincZero** [reply]
可以在他那边修复，见 #66

**Reply by laogao2005** [reply]
好的，谢谢！
